linux/git_mirror

```diff
@@ -813,21 +848,181 @@ function git_push_main()
+                   "$("${GIT}" show-ref --hash "${ref}")" ]; then
+                # No quotes around REGEX
+                if [[ "${ref}" =~ ${GIT_MIRROR_FORCE_PUSH_REFS_REGEX} ]]; then
+                  # NOTE A tag from git_tag_main's should never have to be
+                  # force pushed
+                  "${GIT}" push --force "${repo_url}" "${ref}"
+                  echo "WARNING Fast-forward merge not possible. Force pushed ${ref} to"
+                  echo "    ${repo_url}"
+                  # A ref is tagged by git_tag_main with
+                  # just_git_mirror/{date}/{namespaced_ref}, where namespaced_ref
+                  # is the long-form ref (e.g., refs/heads/master) with the refs/
+                  # domain stripped off
+                  local namespaced_ref="${ref/refs\//}"
+                  # REVIEW This is brittle, but the tag for this ref should
+                  # have already been pushed by the xargs git push above; as
+                  # such, we actually want the second-to-last tag
+                  local tag="$("${GIT}" for-each-ref --sort='-*authordate' \
+                      --format='%(refname:short)' "refs/tags/just_git_mirror/*/${namespaced_ref}" | \
+                      tail -n +2 | head -n 1)"
+                  if [ -n "${tag:+set}" ]; then
+                    # REVIEW Instead of only printing a message here, I could
+                    # create a new tag with this information so it is available
+                    # when fetching these changes in a clone
+                    echo "The original ref for this branch is still referenced by this tag:"
+                    echo "    ${tag}"
+                  else
+                    echo "Could not find the original ref for this branch"
+                  fi
+                else
+                  echo "WARNING Fast-forward merge not possible. ${ref} was not force pushed to"
+                  echo "    ${repo_url}"
+                  echo "because it was not included by GIT_MIRROR_FORCE_PUSH_REFS_REGEX=${GIT_MIRROR_FORCE_PUSH_REFS_REGEX}"
+                fi
+              fi
+            done
+          fi
 
           if git_mirror_has_lfs; then
             next_section "Pushing lfs objects for ${repo_path}"
-            "${GIT}" remote add mirror "${repo_urls[$index]}" 2>/dev/null || "${GIT}" remote set-url mirror "${repo_urls[$index]}"
+            "${GIT}" remote add mirror "${repo_url}" 2>/dev/null || \
+                "${GIT}" remote set-url mirror "${repo_url}"
             # Does not work on file systems, only with real lfs servers, unless you set up lfs-filestore
             "${GIT}" lfs push mirror --all || :
           fi
         popd &> /dev/null
       else
-        echo "No dir found for submodule ${repo_paths[$index]}"
+        echo "No dir found for submodule ${repo_path}"
+      fi
+    done
+  popd &> /dev/null
+
+  # Reset the color
+  echo -n $'\e[0m'
```

Comment:
Why is this necessary?

Reply:
The users prompt was being changed when the call to the function ended